src/util/PageRange.cpp

```diff
@@ -1,76 +1,110 @@
 #include "PageRange.h"
 
-#include <cctype>
-#include <cstdlib>
-
-PageRangeEntry::PageRangeEntry(int first, int last) {
-    this->first = first;
-    this->last = last;
-}
-
-PageRangeEntry::~PageRangeEntry() = default;
-
-auto PageRangeEntry::getLast() const -> int { return this->last; }
-
-auto PageRangeEntry::getFirst() const -> int { return this->first; }
-
-auto PageRange::isSeparator(char c) -> bool { return (c == ',' || c == ';' || c == ':'); }
-
-auto PageRange::parse(const char* str, int pageCount) -> PageRangeVector {
-    PageRangeVector data;
-
-    if (*str == 0) {
-        return data;
+#include <charconv>
+#include <cstddef>
+#include <regex>
+#include <stdexcept>
+#include <string>
+#include <utility>
+
+
+/**
+ * @brief Parse a string of page ranges.
+ *
+ * This function parses a string of page ranges into a vector of pairs of page
+ * numbers. A page range is of the form n, n-, -m, n-m where n, m are positive
+ * integers. The input - is also accepted. Page ranges are separated by `,`, `;`
+ * and `:`. Whitespace is ignored. The parameter 'pageCount' is the largest page
+ * number that may be refered to. The function silently corrects page numbers
+ * equal to 0 or larger than pageCount by truncating to the nearest
+ * acceptable integer.
+ *
+ * Example input and output:
+ *
+ *     parse("1, 2-, -3, 4-5, -, 0-42", 10)
+ *     ===>
+ *     {0, 0}, {1, 9}, {0, 2}, {3, 4}, {0, 9}, {0, 9}
+ *
+ * Note that the page numbers are parsed in the format 1 - pageCount
+ * (permissive) and the return values are in the range 0 - (pageCount-1).
+ *
+ * @param s The string containing page ranges to parse.
+ * @param pageCount The largest page that is possible to refer to.
+ * @exception std::logic_error If `pageCount == 0`.
```

Comment:
Error handling can be performed better by breaking this function into two smaller, private functions:
1. `parse`: Parse the input string into a page range data structure, returning an error message if failing.
2. `validate`: Validate that the page range data structure is valid (e.g., check start <= end, clamp to max page num).

Then a public function should call `parse` and then `validate`, which allows the error handling logic to be cleaner. 

This design also enables the constructor of `PageRange` to be private, so that  `PageRange` can only be constructed through a builder function that validates the provided range.

Reply:
I agree.